update/trigger/trigger.go

```diff
@@ -106,50 +205,68 @@ func (t *trigger) TriggerReceived(originalPayload []byte, data []byte, pkBytes [
 		return true, update.ErrTriggerPubKeyMismatch
 	}
 
-	timestamp, err := t.getTimestampFromPayload(data)
+	arguments, err := t.argumentParser.GetFunctionArguments()
+	if err != nil {
+		return true, err
+	}
+
+	if len(arguments) != 2 {
+		return true, update.ErrIncorrectHardforkMessage
+	}
+
+	timestamp, err := t.getIntFromArgument(string(arguments[0]))
 	if err != nil {
 		return true, err
 	}
 
 	currentTimeStamp := t.getTimestampHandler()
 	if timestamp+int64(hardforkGracePeriod.Seconds()) < currentTimeStamp {
-		return true, fmt.Errorf("%w out of grace period message", update.ErrIncorrectHardforkMessage)
+		return true, fmt.Errorf("%w timestamp out of grace period message", update.ErrIncorrectHardforkMessage)
 	}
 
-	return true, t.doTrigger(originalPayload)
-}
+	epoch, err := t.getIntFromArgument(string(arguments[1]))
+	if err != nil {
+		return true, err
+	}
 
-func (t *trigger) getTimestampFromPayload(payload []byte) (int64, error) {
-	splits := bytes.Split(payload, []byte(dataSeparator))
-	if len(splits) < 2 {
-		return 0, fmt.Errorf("%w, not enough tokens found: expected 2, found: %d",
-			update.ErrIncorrectHardforkMessage,
-			len(splits),
-		)
+	currentEpoch := int64(t.epochProvider.MetaEpoch())
+	if currentEpoch-epoch > epochGracePeriod {
+		return true, fmt.Errorf("%w epoch out of grace perdiod", update.ErrIncorrectHardforkMessage)
 	}
 
-	encodedValue := string(splits[1])
-	n, err := strconv.ParseInt(encodedValue, 10, 64)
+	t.mutTriggered.Lock()
+	t.triggered = true
+	t.recordedTriggerMessage = originalPayload
+	t.epoch = uint32(epoch)
+	t.mutTriggered.Unlock()
+
+	t.doTrigger(uint32(epoch))
+
+	return true, nil
+}
+
+func (t *trigger) getIntFromArgument(value string) (int64, error) {
+	n, err := strconv.ParseInt(value, 10, 64)
 	if err != nil {
 		return 0, fmt.Errorf("%w, convert error, `%s` is not a valid int",
 			update.ErrIncorrectHardforkMessage,
-			encodedValue,
+			value,
 		)
 	}
 
 	return n, nil
 }
 
-func (t *trigger) callAddedDataHandlers() {
+func (t *trigger) callAddedDataHandlers(epoch uint32) {
 	t.mutTriggerHandlers.RLock()
 	for _, handler := range t.triggerHandlers {
-		go handler()
+		go handler(epoch)
 	}
 	t.mutTriggerHandlers.RUnlock()
 }
 
 // RegisterHandler will add a trigger event handler to the list
-func (t *trigger) RegisterHandler(handler func()) error {
+func (t *trigger) RegisterHandler(handler func(epoch uint32)) error {
```

Comment:
do we still need these? (RegisterHandler, callAddedDataHandlers, associated mutex and triggerHandlers)?

Reply:
maybe not.

Reply:
Removed in the next PR